sunkit_instruments/suvi/suvi.py

```diff
@@ -0,0 +1,820 @@
+        for a 171 L1b file. If this pattern is not found in the
+        filename, the file will not be recognized.
+
+    .. note::
+        The despiking relies on the presence of the data quality
+        flags (DQF) in the first extension of a SUVI L1b FITS file.
+        Early in the mission, the DQF extension was not present
+        yet, so the despiking cannot be done with this function
+        for those early files.
+
+    Parameters
+    ----------
+    the_input: `str` or `tuple` of 2-D `~numpy.ndarray` containing (data, DQF)
+        Filename or tuple of (data, DQF) to despike.
+
+    filter_width: `int`, optional. Default: 7.
+        The filter width for the Gaussian filter. If NaNs are still
+        present in the despiked image, try increasing this value.
+
+    return_map: `bool`, optional. Default: False.
+        If True, returns a `~sunpy.map.Map` instead of a `~numpy.ndarray`.
+        Ignored if the_input is not a filename string.
+
+    Returns
+    -------
+    despiked_image: `~numpy.ndarray`
+        The despiked L1b image as a numpy array.
+
+    map: `~sunpy.map.Map` if return_map=True.
+        The despiked L1b image as a `~sunpy.map.Map`.
+    """
+
+    if isinstance(the_input, tuple):
+        image = the_input[0]
+        dqf_mask = the_input[1]
+    elif isinstance(the_input, str):
+        L1b_matches = ['-L1b-Fe093', '-L1b-Fe131', '-L1b-Fe171', '-L1b-Fe195', '-L1b-Fe284', '-L1b-He303']
+        if any(fn in os.path.basename(the_input) for fn in L1b_matches):
+            header, image, dqf_mask = sunkit_instruments.suvi.read_suvi(the_input, return_DQF=True)
+        else:
+            raise ValueError("File "+the_input+" does not look like a SUVI L1b file.")
+    else:
+        raise TypeError("Input must be string or tuple of 2-D arrays.")
+
+    image_with_nans = numpy.copy(image)
+    image_with_nans[numpy.where(dqf_mask == 4)] = numpy.nan
+    indices = numpy.where(numpy.isnan(image_with_nans))
+    image_gaussian_filtered = gaussian_filter(image, filter_width)
+    despiked_image = numpy.copy(image_with_nans)
+    despiked_image[indices] = image_gaussian_filtered[indices]
+
+    if return_map and isinstance(the_input, str):
+        return sunpy.map.Map(despiked_image, header)
+    else:
+        return despiked_image
+
+
+# helper function: parser for the SUVI websites using BeautifulSoup
+def _list_url_directory(url, ext=''):
+    page = requests.get(url).text
+    soup = BeautifulSoup(page, 'html.parser')
+    return [url + node.get('href') for node in soup.find_all('a') if node.get('href').endswith(ext)]
+
+
+def download_data_from_NOAA(date_time, spacecraft=16, wavelength=171, outdir=None,
+                            composites=False, query_only=False, verbose=False):
+    """
+    Download SUVI L1b or L2 HDR composite FITS files from the NOAA websites directly.
+    Intended as an alternative to getting it from the VSO with fido. For L1b
```

Comment:
We want any download functionality to go through Fido in sunpy. This means this function can not be accepted in its current state. 

You would need to create a new Fido client that would allow you to use the URL to scrap much like your current function does. But with the client code we have in sunpy, it will handle most of that for you automatically. Please see https://docs.sunpy.org/en/latest/dev_guide/contents/extending_fido.html on "Writing a new “scraper” client".
This would also need to go into a new `net.py` or `client.py` file as well.

This guide is quite new and untested by people who aren't sunpy developers, so if you have issues or anything is unclear, please do say so and we can work on fixing that.

Reply:
Ok, if this is the policy, I will remove this function entirely. The whole point was to give the user an alternative to Fido.

Reply:
Ah in that case, personally I would want it removed. It would be really good to have it as a Fido Client so users can get these files through fido if these files are different to the files that the VSO serve. 

Are there any issues with fido that you are having and wanted to bypass it?

Reply:
SUVI is already in the VSO, which is what I think Fido uses. Ultimately the SUVI files get downloaded from the NOAA website anyway, but there have been some issues with the VSO lately, so I just wanted to remove this (possibly temporarily not working) layer and allow access to the NOAA website directly. I get the whole "we want our code to be consitent" policy, but does it really hurt anyone to allow this function? Either way, let me know if you want it removed or not.

Reply:
I will check with the others on their thoughts.

For me personally, we have spent quite a bit of time on getting Fido into a state that allows people to write custom clients that does most of what this function does but allows users to use Fido to search and fetch files instead using a custom download function. 

What "official" policy we do have (I am unsure if it was written down) was that we don't want to duplicate the VSO so we try to limit our clients without a solid reason. But I am happy to add this client if need be.

Reply:
Ok, thanks for the explanation. Maybe I will look into writing a new scraper with Fido, depends how complicated it is. I just had this one already and thought I would include it in case anyone has the need to use it, but it's not the hill I want to die on :-) Let me know what you and the others think and I will leave it or remove it.